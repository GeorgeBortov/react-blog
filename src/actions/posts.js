import database from '../firebase/firebase';

// ADD_POST
export const addPost = (post) => ({
    type: 'ADD_POST',
    post
});

export const startAddPost = (postData = {}) => {
    return (dispatch, getState) => {
        const authorID = getState().auth.uid;
        const {
            title = '',
            body = '',
            createdAt = 0,
            authorName = '',
        } = postData;
        const post = { title, body, createdAt, authorName, authorID }
        
        return database.ref(`posts`).push(post).then((ref) => {
            dispatch(addPost({
                id: ref.key,
                ...post
            }));
        });
    }
};

// REMOVE_EXPENSE
export const removePost = ({ id } = {}) => ({
    type: 'REMOVE_POST',
    id
});

export const startRemovePost = ({ id } = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`posts/${id}`).remove().then(() => {
            dispatch(removePost({ id }));
        });
    }
};

// EDIT_EXPENSE
export const editPost = (id, updates) => ({
    type: 'EDIT_POST',
    id,
    updates
});

export const startEditPost = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`posts/${id}`).update(updates).then(() => {
            dispatch(editPost(id, updates));
        });
    }
};

// SET_POSTS
export const setPosts = (posts) => ({
    type: 'SET_POSTS',
    posts
});

export const startSetPosts = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        // return database.ref(`users/${uid}/posts`)
        return database.ref(`posts`)
        .once('value')
        .then((snapshot) => {
            const posts = [];

            snapshot.forEach((childSnapshot) => {
                posts.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            })
            dispatch(setPosts(posts));
        });
    }
};