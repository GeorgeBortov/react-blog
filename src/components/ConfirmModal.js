import React from 'react';
import Modal from 'react-modal';

const ConfirmModal = ({isOpen, onRequestClose, onRemove, postTitle}) => (
    <Modal
        isOpen={isOpen}
        onRequestClose={onRequestClose}
        contentLabel="Confirmation Modal"
        closeTimeoutMS={200}
        appElement={document.getElementById('app')}
        className={'modal'}
    >
        <div className="modal__body">
            <button onClick={onRequestClose} className="modal__close"></button>
            <h3 className="modal__title">Are you sure you want to delete a "{postTitle}" post?</h3>
            <button onClick={onRemove} className="button removeExpense">Yes</button>
            <button onClick={onRequestClose} className="button button--secondary">No</button>
        </div>
               
    </Modal>
);

export default ConfirmModal;