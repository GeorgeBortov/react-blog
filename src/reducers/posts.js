// Posts Reducer

const postsReducerDefaultState = {
    postsArr: [],
    comentsArr: [],
    postTotalNumb: 0
};

export default (state = postsReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_POST':
            return {
                ...state,
                postsArr: [
                    ...state.postsArr,
                    action.post
                ]
            };
        case 'ADD_COMMENT':
            return {
                ...state,
                comentsArr: [
                    ...state.comentsArr,
                    action.comment
                ]
                // postsArr: state.postsArr.comments.concat(action.comment)
                
            };
        case 'SET_COMMENTS':
            return {
                ...state,
                comentsArr: action.comments, 
            };
        case 'REMOVE_POST':
            return {
                ...state,
                postsArr: state.postsArr.filter(({ id }) => id !== action.id)
            };
        case 'EDIT_POST':
            return {
                ...state,
                postsArr: state.postsArr.map((post) => {
                    if (post.id === action.id) {
                        return {
                            ...post,
                            ...action.updates
                        }
                    } else {
                        return post;
                    }
                })
            };
        case 'SET_POSTS':
            return {
                ...state,
                postsArr: action.posts,
                postTotalNumb: action.postsNumb
            };
        default:
            return state;
    }
};
