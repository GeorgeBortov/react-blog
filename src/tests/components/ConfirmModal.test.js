import React from 'react';
import { shallow } from 'enzyme';
import ConfirmModal from '../../components/ConfirmModal';
import posts from '../fixtures/posts';

let onRequestClose, onRemove;

beforeEach(() => {
    onRequestClose = jest.fn();
    onRemove = jest.fn();
});

test('should render close ConfirmModal correctly', () => {
    const wrapper = shallow(<ConfirmModal
        postTitle={posts[0].title}
        isOpen={false}
        onRemove={onRemove}
        onRequestClose={onRequestClose}
    />);
    expect(wrapper).toMatchSnapshot();
});

test('should render open ConfirmModal correctly', () => {
    const wrapper = shallow(<ConfirmModal
        postTitle={posts[0].title}
        isOpen={true}
        onRemove={onRemove}
        onRequestClose={onRequestClose}
    />);
    expect(wrapper).toMatchSnapshot();
});

test('should handle onRemove', () => {
    const wrapper = shallow(<ConfirmModal
        onRemove={onRemove}
    />);
    wrapper.find('.removeExpense').simulate('click');
    expect(onRemove).toHaveBeenCalled();
});

test('should handle onRequestClose with button "NO" ', () => {
    const wrapper = shallow(<ConfirmModal
        onRequestClose={onRequestClose}
    />);
    wrapper.find('.button--secondary').simulate('click');
    expect(onRequestClose).toHaveBeenCalled();
});

test('should handle onRequestClose with button "X" ', () => {
    const wrapper = shallow(<ConfirmModal
        onRequestClose={onRequestClose}
    />);
    wrapper.find('.modal__close').simulate('click');
    expect(onRequestClose).toHaveBeenCalled();
});