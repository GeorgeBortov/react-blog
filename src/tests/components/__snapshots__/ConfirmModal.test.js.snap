// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render close ConfirmModal correctly 1`] = `
<Modal
  appElement={null}
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  className="modal"
  closeTimeoutMS={200}
  contentLabel="Confirmation Modal"
  isOpen={false}
  onRequestClose={[MockFunction]}
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
>
  <div
    className="modal__body"
  >
    <button
      className="modal__close"
      onClick={[MockFunction]}
    />
    <h3
      className="modal__title"
    >
      Are you sure you want to delete a "
      test title 1
      " post?
    </h3>
    <button
      className="button removeExpense"
      onClick={[MockFunction]}
    >
      Yes
    </button>
    <button
      className="button button--secondary"
      onClick={[MockFunction]}
    >
      No
    </button>
  </div>
</Modal>
`;

exports[`should render open ConfirmModal correctly 1`] = `
<Modal
  appElement={null}
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  className="modal"
  closeTimeoutMS={200}
  contentLabel="Confirmation Modal"
  isOpen={true}
  onRequestClose={[MockFunction]}
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
>
  <div
    className="modal__body"
  >
    <button
      className="modal__close"
      onClick={[MockFunction]}
    />
    <h3
      className="modal__title"
    >
      Are you sure you want to delete a "
      test title 1
      " post?
    </h3>
    <button
      className="button removeExpense"
      onClick={[MockFunction]}
    >
      Yes
    </button>
    <button
      className="button button--secondary"
      onClick={[MockFunction]}
    >
      No
    </button>
  </div>
</Modal>
`;
